core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/SplittingTransformer.java

```diff
@@ -0,0 +1,335 @@
+public class SplittingTransformer<ResponseT, ResultT> implements SdkPublisher<AsyncResponseTransformer<ResponseT, ResponseT>> {
+    // Note: comments in this class that start with '§' reference a rule from the reactive stream specification.
+
+    private static final Logger log = Logger.loggerFor(SplittingTransformer.class);
+
+    /**
+     * The AsyncResponseTransformer on which the {@link AsyncResponseTransformer#split(long) split} method was called.
+     */
+    private final AsyncResponseTransformer<ResponseT, ResultT> upstreamResponseTransformer;
+
+    /**
+     * Set to true once {@code .prepare()} is called on the upstreamResponseTransformer
+     */
+    private final AtomicBoolean preparedCalled = new AtomicBoolean(false);
+
+    /**
+     * Set to true once {@code .onResponse()} is called on the upstreamResponseTransformer
+     */
+    private final AtomicBoolean onResponseCalled = new AtomicBoolean(false);
+
+    /**
+     * Set to true once {@code .onStream()} is called on the upstreamResponseTransformer
+     */
+    private final AtomicBoolean onStreamCalled = new AtomicBoolean(false);
+
+    /**
+     * Set to true once {@code .concel()} is called in the subscription of the downstream subscriber
+     */
+    private final AtomicBoolean isCancelled = new AtomicBoolean(false);
+
+    /**
+     * Future to track the status of the upstreamResponseTransformer. Will be completed when the future returned by calling
+     * {@code prepare()} on the upstreamResponseTransformer itself completes.
+     */
+    private final CompletableFuture<ResultT> returnFuture;
+
+    /**
+     * The buffer size used to buffer the content received from the downstream subscriber
+     */
+    private final int maximumBufferSize;
+
+    /**
+     * This publisher is used to send the bytes received from the downstream subscriber's transformers to a
+     * {@link DelegatingBufferingSubscriber} that will buffer a number of bytes specified by the {@code maximumBufferSize}.
+     */
+    private final SimplePublisher<ByteBuffer> publisherToUpstream = new SimplePublisher<>();
+
+    /**
+     * The downstream subscriber that is subscribed to this publisher.
+     */
+    private Subscriber<? super AsyncResponseTransformer<ResponseT, ResponseT>> downstreamSubscriber;
+
+    /**
+     * The amount requested by the downstream subscriber that is still left to fulfill. Updated. when the
+     * {@link Subscription#request(long) request} method is called on the downstream subscriber's subscription. Corresponds to the
+     * number of {@code AsyncResponseTransformer} that will be published to the downstream subscriber.
+     */
+    private final AtomicLong outstandingDemand = new AtomicLong(0);
+
+    /**
+     * tracks if the publisher is currently sending AsyncResponseTransformer to the downstreamSubscriber
+     */
+    private final AtomicBoolean emitting = new AtomicBoolean(false);
+
+    // 'boundedness' info
+    private final long maxElements;
```

Comment:
Is there any reason we need `maxElements`?

Reply:
To be compliant with **[spec 3.2](https://github.com/reactive-streams/reactive-streams-jvm/blob/v1.0.4/README.md#3.2)**. It seems publisher need to have some kind of upper bound

Reply:
we have an on going discussion about this

Reply:
It seems the failed test expects the publisher to publish bounded number, so suggesting the following as a workaround. This way, we can remove maxElements. I'd say the implementatoin is compliant; it's just that the test case is not supported in our case.

```java
    @Override
    public Publisher<AsyncResponseTransformer<Object, Object>> createPublisher(long l) {
        CompletableFuture<ResponseBytes<Object>> future = new CompletableFuture<>();
        AsyncResponseTransformer<Object, ResponseBytes<Object>> upstreamTransformer = AsyncResponseTransformer.toBytes();
        SplittingTransformer.Builder<Object, ResponseBytes<Object>> builder =
            SplittingTransformer.<Object, ResponseBytes<Object>>builder()
                                .upstreamResponseTransformer(upstreamTransformer)
                                .bufferSize(64 * 1024)
                                .returnFuture(future);
        return SdkPublisher.adapt(builder.build()).limit(6);
    }

    @Override
    public long maxElementsFromPublisher() {
        return Long.MAX_VALUE;
    }
```

Reply:
maxElements removed, thanks